### DISCUSSION TITLE
Understanding index column in a `StdDataFrame<I>` and templating

### DISCUSSION THREAD

**Discussion by dereckmezquita** [hosseinmoein/DataFrame#287] [Q&A] [answered]
One thing I've not understood from the docs is why template the index type for a `DataFrame`? I can understand using certain types for the index column such as `int`/`uin64_t`/`DateTime` etc. It makes sense to me that these should be unique values (`DateTime` might not).

However it is also valid to create a `StdDataFrame<bool>`.  Can such a construct have a use?

```cpp
std::vector<double> dblvec2 = { 0.998, 0.3456, 0.056, 0.15678, 0.00345, 0.923, 0.06743, 0.1 };
std::vector<std::string> strvec = { "Col_name", "Col_name", "Col_name", "Col_name", "Col_name" };

StdDataFrame<bool> df2;
df2.load_data(std::move(idx_bool),
    std::make_pair("dbl_col_2", dblvec2),
    std::make_pair("str_col", strvec));
```

I just want to better understand some of the reasoning/use; the why behind the way `DataFrame` was built and the data structures behind it.

I'm in the process of learning so thank you.

**Comment by hosseinmoein** [comment, marked as answer]
An index column in a DataFrame is unlike an index in a SQL table. SQL table index makes access efficient. It doesn’t give you any more information. The index column in a DataFrame is metadata about the data in the DataFrame. Each entry in the index describes the given row. It could be time, frequency, …, or a set of descriptors in a struct (like temperature, altitude, …). So, there can be almost infinite possibilities. That’s why I kept the index column as a template which also differentiates the DataFrame type.
Whether a bool index column type is relevant or not is unimportant.
